gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/CaResources.java

```diff
@@ -74,20 +75,44 @@ private void importCertificates(Path jdkInstallationDirectory, List<Certificate>
             char[] passwd = "changeit".toCharArray();
             Path jksPath = jdkInstallationDirectory.resolve("lib/security/cacerts");
             KeyStore jks = loadKeystore(passwd, jksPath);
-            for (Certificate cert : certificates) {
-                String alias = getAlias((X509Certificate) cert);
+            Set<BigInteger> existingSerialNumbers = getExistingCertificates(jks);
+            logger.log(String.format("Certificate %s imported successfully into the KeyStore.", existingSerialNumbers));
+            List<X509Certificate> newCertificates = certificates.stream()
+                    .map(X509Certificate.class::cast)
+                    .filter(certificate -> !existingSerialNumbers.contains(certificate.getSerialNumber()))
+                    .collect(Collectors.toList());
+            for (X509Certificate certificate : newCertificates) {
+                String alias = getAlias(certificate);
                 logger.log(String.format("Certificate %s imported successfully into the KeyStore.", alias));
-                jks.setCertificateEntry(alias, cert);
+                jks.setCertificateEntry(alias, certificate);
             }
-
             jks.store(new BufferedOutputStream(new FileOutputStream(jksPath.toFile())), passwd);
-
-            logger.log("Certificates imported successfully into the KeyStore.");
         } catch (KeyStoreException | CertificateException | IOException | NoSuchAlgorithmException e) {
             throw new RuntimeException("Failed to import certificates", e);
         }
     }
 
+    private static Set<BigInteger> getExistingCertificates(KeyStore keyStore) {
+        try {
+            return Collections.list(keyStore.aliases()).stream()
+                    .map(alias -> {
+                        try {
+                            return Optional.ofNullable(keyStore.getCertificate(alias))
+                                    .map(X509Certificate.class::cast);
+                        } catch (KeyStoreException e) {
+                            throw new RuntimeException(e);
+                        } catch (ClassCastException e) {
```

Comment:
Rather than catching the exception, can we do `.filter(X509Certificate.class::isAssignableFrom)` before the `map(cast)`?

Reply:
If you use streamex, it becomes even nicer:

```java
try {
  StreamEx.of(keyStore.aliases())
    // might need to extract to a method or inline inside, but it's minimal
    .map(keystore::getCertificate) 
    .nonNull()
    .select(X509Certificate.class)
    .map(X509Certificate::getSerialNumber)
    .toSet();
} catch (KeyStoreException e) {
  throw new RuntimeException(e);
}
```

Reply:
I haven't used StreamEx because CaResources will be included in the `gradle-jdks-setup` fat jar that will be shipped to all repos, so I am trying to keep the jar dependencies as small as possible: https://github.com/palantir/gradle-jdks/blob/develop/gradle-jdks-setup/build.gradle#L11